### DISCUSSION TITLE
"NumChannels" of get_nodeattr()

### DISCUSSION THREAD

**Discussion by caoimhelynch** [Xilinx/finn#551] [Q&A] [answered]
Trying to follow the cnv_end2end_example, the function to_hls.InferThresholdingLayer() takes the first index array of the multi-threshold which is [3] (this is also the original number of channels for my model). However this 3 is compared with the NumChannels which for some reason is returned by get_nodeattr() as 150 (which is the image size). These are of course not the same and causes the line to fall over with error AssertionError: Channels of threshold matrix are not as expected (mh).
Would there be any suggestions about how I am passing my model in to get the wrong NumChannel back? Thanks

**Comment by B-Willems** [comment, marked as answer]
did you ever find a solution to this problem? I am encountering the same issue with my convolutional Neural Network.

**Reply by fpjentzsch** [reply]
Hi, could either of you share the .onnx model before you apply `InferThresholdingLayer()` please?

**Reply by B-Willems** [reply]
In the meantime I found a solution to my problem. I used a permute action on my input tensor to make it match the dimentions needed by my network. However this caused the problem as far as i know. When i moved the permute outside of the model. Aka I clean up my dataset, and write it as a clean dataset with the right dimentions (1, 4, 78, 78) instead of (1, 78, 78, 4) the error is solved. I presume this is because the Transpose layers in FINN don't reshape my tensor correctly when they need to represent a permute layer. Anyway that's how i understand it :p.

**Reply by B-Willems** [reply]
In short, the input of your PyTorch/Brevitas model should be (x, y, z, q) where x is your batchsize, y is your amount of channels (for example 3 when working with RGB), and z, q are the dimentions of your input image.

**Reply by caoimhelynch** [reply]
I believe this was the solution I found also to this issue. I required a permute for the training of my model to ensure shape (1, 150, 150, 3), however this returned 150 channels when processed by Finn, giving the error above. Removing the permute to create the ONNX model gave me the correct dimensions for Finn (1, 3, 150, 150) in my case.
